File: genice_julia/genice_core.jl
# genice-core: ice rule を満たす有向グラフを生成（Julia 実装）
# 入出力は配列のみ。adj は 1-based: adj[i] = ノード i-1 の隣接（0-based ノード番号の Vector）

using Random

# ---------- graph_arrays 相当 ----------
node_to_idx(v, n) = v >= 0 ? v : n + (-1 - v)
idx_to_node(i, n) = i <= n ? i : -1 - (i - n)

# 0-based ノード v の隣接を取得（adj は 1-based）
adj_at(adj, v) = adj[v + 1]

function connected_components(n_nodes, adj)
    seen = falses(n_nodes)
    components = Vector{Int}[]
    for v in 0:(n_nodes - 1)
        if !seen[v + 1]
            comp = Int[]
            stack = [v]
            seen[v + 1] = true
            while !isempty(stack)
                cur = pop!(stack)
                push!(comp, cur)
                for w in adj_at(adj, cur)
                    if !seen[w + 1]
                        seen[w + 1] = true
                        push!(stack, w)
                    end
                end
            end
            push!(components, comp)
        end
    end
    components
end

# ---------- topology: trace_path, find_path ----------
function trace_path!(n_nodes, adj, path, vs)
    while true
        last_, head = path[end-1], path[end]
        next_node = -1
        for w in adj_at(adj, head)
            if w in vs && w != last_
                next_node = w
                break
            end
        end
        next_node < 0 && return
        push!(path, next_node)
        next_node == path[1] && return
    end
end

function find_path(n_nodes, adj, vertex_set)
    isempty(vertex_set) && return Int[]
    vs = Set(vertex_set)
    head = vertex_set[1]
    neighbors = [w for w in adj_at(adj, head) if w in vs]
    isempty(neighbors) && return Int[]
    if length(neighbors) == 1
        path = [head, neighbors[1]]
        trace_path!(n_nodes, adj, path, vs)
        return path
    end
    c0 = [head, neighbors[1]]
    trace_path!(n_nodes, adj, c0, vs)
    c0[end] == head && return c0
    c1 = [head, neighbors[2]]
    trace_path!(n_nodes, adj, c1, vs)
    [reverse(c0); c1[2:end]]
end

# ---------- topology: _divide, noodlize ----------
function divide!(n_nodes, adj, vertex, offset)
    nei = copy(adj_at(adj, vertex))
    while length(nei) < 4
        push!(nei, -1)
    end
    nei = nei[1:4]
    valid = [x for x in nei if x >= 0]
    length(valid) < 2 && return
    shuffle!(valid)
    migrants = [valid[1], valid[2]]
    new_vertex = vertex + offset
    for m in migrants
        a = adj_at(adj, m)
        i = findfirst(==(vertex), a)
        isnothing(i) || deleteat!(a, i)
        a = adj_at(adj, vertex)
        i = findfirst(==(m), a)
        isnothing(i) || deleteat!(a, i)
        push!(adj_at(adj, new_vertex), m)
        push!(adj_at(adj, m), new_vertex)
    end
end

function noodlize(n_orig, adj, fixed_out, fixed_in)
    n_nodes = 2 * n_orig
    adj_noodles = [copy(adj_at(adj, v)) for v in 0:(n_orig-1)]
    append!(adj_noodles, [Int[] for _ in 1:n_orig])
    for u in 0:(n_orig-1)
        iu = node_to_idx(u, n_orig)
        for v in fixed_out[iu + 1]
            if 0 <= v < n_orig
                anu = adj_noodles[u + 1]
                anv = adj_noodles[v + 1]
                i = findfirst(==(v), anu)
                isnothing(i) || deleteat!(anu, i)
                i = findfirst(==(u), anv)
                isnothing(i) || deleteat!(anv, i)
            end
        end
    end
    for v in 0:(n_orig-1)
        i = node_to_idx(v, n_orig)
        nfixed = length(fixed_out[i + 1]) + length(fixed_in[i + 1])
        nfixed == 0 && divide!(n_nodes, adj_noodles, v, n_orig)
    end
    (n_nodes, adj_noodles)
end

# ---------- topology: _decompose_complex_path ----------
function decompose_complex_path(path)
    isempty(path) && return Int[]
    order = Dict{Int,Int}()
    store = Int[]
    order[path[1]] = 0
    push!(store, path[1])
    result = Vector{Int}[]
    headp = 2
    while headp <= length(path)
        node = path[headp]
        if haskey(order, node)
            size = length(order) - order[node]
            cycle = [store[end-size+1:end]; node]
            push!(result, cycle)
            for v in cycle[2:end]
                delete!(order, v)
            end
            store = store[1:end-size]
        end
        order[node] = length(order)
        push!(store, node)
        headp += 1
    end
    length(store) > 1 && push!(result, store)
    result
end

function split_into_simple_paths(n_orig, n_nodes, adj)
    components = connected_components(n_nodes, adj)
    result = Vector{Int}[]
    for vertex_set in components
        path = find_path(n_nodes, adj, vertex_set)
        flatten = [v % n_orig for v in path]
        append!(result, decompose_complex_path(flatten))
    end
    result
end

# ---------- connect_matching_paths 用ヘルパ ----------
function copy_directed(n_orig, out_adj, in_adj)
    size = n_orig + 4
    ([copy(out_adj[i]) for i in 1:size], [copy(in_adj[i]) for i in 1:size])
end

function get_perimeters(n_orig, adj, out_adj, in_adj)
    in_peri = Set{Int}()
    out_peri = Set{Int}()
    for i in 1:n_orig
        node = i - 1
        deg_g = length(adj_at(adj, node))
        in_d = length(in_adj[i])
        out_d = length(out_adj[i])
        in_d + out_d >= deg_g && continue
        in_d > out_d && push!(out_peri, node)
        in_d < out_d && push!(in_peri, node)
    end
    (in_peri, out_peri)
end

function choose_free_edge(n_orig, adj, out_adj, in_adj, node)
    neis = copy(adj_at(adj, node))
    while length(neis) < 4
        push!(neis, -1 - length(neis))
    end
    neis = neis[1:4]
    shuffle!(neis)
    for nei in neis
        nei == -1 && continue
        i_node = node_to_idx(node, n_orig)
        i_nei = nei >= 0 ? node_to_idx(nei, n_orig) : n_orig + (-1 - nei)
        has_out = nei in out_adj[i_node + 1]
        has_in = node in out_adj[i_nei + 1]
        !has_out && !has_in && return nei
    end
    -999
end

function remove_dummy_nodes!(n_orig, out_adj, in_adj)
    for i in (n_orig+1):(n_orig+4)
        empty!(out_adj[i])
        empty!(in_adj[i])
    end
    for v in 0:(n_orig-1)
        out_adj[v+1] = [w for w in out_adj[v+1] if 0 <= w < n_orig]
        in_adj[v+1] = [u for u in in_adj[v+1] if 0 <= u < n_orig]
    end
end

function connect_matching_paths(n_orig, adj, fixed_out, fixed_in)
    (f_out, f_in) = copy_directed(n_orig, fixed_out, fixed_in)
    (in_peri, out_peri) = get_perimeters(n_orig, adj, f_out, f_in)
    derived_cycles = Vector{Int}[]

    add_edge(u, v) = begin
        iu = node_to_idx(u, n_orig)
        iv = node_to_idx(v, n_orig)
        push!(f_out[iu + 1], v)
        push!(f_in[iv + 1], u)
    end
    in_degree(node) = length(f_in[node_to_idx(node, n_orig) + 1])
    out_degree(node) = length(f_out[node_to_idx(node, n_orig) + 1])

    while !isempty(out_peri)
        node = first(out_peri)
        delete!(out_peri, node)
        path = [node]
        cur = node
        while true
            cur < 0 && break
            cur in in_peri && (delete!(in_peri, cur); break)
            out_degree(cur) * 2 > 4 && return nothing
            length(adj_at(adj, cur)) == out_degree(cur) + in_degree(cur) && return nothing
            next_node = choose_free_edge(n_orig, adj, f_out, f_in, cur)
            next_node == -999 && return nothing
            add_edge(cur, next_node)
            if next_node >= 0
                push!(path, next_node)
                in_degree(cur) > out_degree(cur) && push!(out_peri, cur)
            end
            cur = next_node
            idx = findfirst(==(cur), path[1:end-1])
            if !isnothing(idx)
                push!(derived_cycles, path[idx:end])
                path = path[1:idx]
            end
        end
    end

    while !isempty(in_peri)
        node = first(in_peri)
        delete!(in_peri, node)
        path = [node]
        cur = node
        while true
            cur < 0 && break
            cur in out_peri && (delete!(out_peri, cur); break)
            out_degree(cur) * 2 > 4 && return nothing
            length(adj_at(adj, cur)) == out_degree(cur) + in_degree(cur) && return nothing
            next_node = choose_free_edge(n_orig, adj, f_out, f_in, cur)
            next_node == -999 && return nothing
            if next_node >= 0
                push!(path, next_node)
            end
            add_edge(next_node, cur)
            next_node >= 0 && in_degree(cur) < out_degree(cur) && push!(in_peri, cur)
            cur = next_node
            idx = findfirst(==(cur), path[1:end-1])
            if !isnothing(idx)
                push!(derived_cycles, path[idx:end])
                path = path[1:idx]
            end
        end
    end

    remove_dummy_nodes!(n_orig, f_out, f_in)
    (f_out, f_in, derived_cycles)
end

# ---------- directed edges ----------
function arrays_to_directed_edges(n_orig, out_adj, in_adj)
    edges = Tuple{Int,Int}[]
    for i in 1:length(out_adj)
        u = idx_to_node(i - 1, n_orig)
        u < 0 && continue
        for v in out_adj[i]
            (0 <= v < n_orig) && push!(edges, (u, v))
        end
    end
    edges
end

path_edges(paths) = [(p[i], p[i+1]) for p in paths for i in 1:length(p)-1]

function finally_fixed_edges(n_orig, fixed_out, fixed_in, derived_cycles)
    edges = arrays_to_directed_edges(n_orig, fixed_out, fixed_in)
    cycle_set = Set([(c[i], c[i+1]) for c in derived_cycles for i in 1:length(c)-1])
    [(u, v) for (u, v) in edges if (u, v) ∉ cycle_set]
end

# ---------- main ----------
"""
    ice_graph_edges(n_orig, adj, fixed_out, fixed_in; pairing_attempts=100)

Returns `Vector{Tuple{Int,Int}}` of directed edges, or `nothing` if no solution.
adj, fixed_out, fixed_in are 1-based: adj[i] = node i-1's data.
"""
function ice_graph_edges(n_orig, adj, fixed_out, fixed_in; pairing_attempts=100)
    f_out = fixed_out
    f_in = fixed_in
    derived_cycles = Vector{Int}[]

    n_fixed = sum(length, f_out)
    solved = false
    if n_fixed > 0
        for attempt in 1:pairing_attempts
            result = connect_matching_paths(n_orig, adj, f_out, f_in)
            if !isnothing(result)
                (f_out, f_in, derived_cycles) = result
                solved = true
                break
            end
        end
        !solved && return nothing
    end

    fin_fixed = finally_fixed_edges(n_orig, f_out, f_in, derived_cycles)
    (n_nodes, adj_noodles) = noodlize(n_orig, adj, f_out, f_in)
    paths = split_into_simple_paths(n_orig, n_nodes, adj_noodles)
    append!(paths, derived_cycles)

    all_edges = [fin_fixed; path_edges(paths)]
    all_edges
end

# ---------- 使用例 ----------
if abspath(PROGRAM_FILE) == @__FILE__
    Random.seed!(42)
    # adj[i] = ノード i-1 の隣接（0-based 番号）
    adj = [[1, 2], [0, 2], [0, 1], [4], [3]]
    n_orig = 5
    fixed_out = [Int[] for _ in 1:(n_orig+4)]
    fixed_in = [Int[] for _ in 1:(n_orig+4)]

    result = ice_graph_edges(n_orig, adj, fixed_out, fixed_in)
    if !isnothing(result)
        println("edges: ", length(result))
        println("sample: ", result[1:min(5, end)])
    else
        println("no solution")
    end
end

File: genice_nim/genice_core.nim
## genice-core: ice rule を満たす有向グラフを生成（Nim 実装）
## 入出力は配列のみ。API: iceGraphEdges(nOrig, adj, fixedOut, fixedIn, ...) -> seq[(int,int)]

import std/algorithm
import std/options
import std/random
import std/sequtils
import std/sets
import std/tables

type
  Adj* = seq[seq[int]]
  DirectedEdges* = seq[(int, int)]

# ---------- graph_arrays 相当 ----------
func nodeToIdx(v, n: int): int = (if v >= 0: v else: n + (-1 - v))
func idxToNode(i, n: int): int = (if i < n: i else: -1 - (i - n))

func connectedComponents(nNodes: int; adj: Adj): seq[seq[int]] =
  var seen = newSeq[bool](nNodes)
  for v in 0 ..< nNodes:
    if not seen[v]:
      var comp: seq[int]
      var stack = @[v]
      seen[v] = true
      while stack.len > 0:
        let cur = stack.pop
        comp.add cur
        for w in adj[cur]:
          if not seen[w]:
            seen[w] = true
            stack.add w
      result.add comp

# ---------- topology: trace_path, find_path ----------
func tracePath(nNodes: int; adj: Adj; path: var seq[int]; vertexSet: HashSet[int]): void =
  let vs = vertexSet
  while true:
    let last = path[^2]
    let head = path[^1]
    var nextNode = -1
    for w in adj[head]:
      if w in vs and w != last:
        nextNode = w
        break
    if nextNode < 0:
      return
    path.add nextNode
    if nextNode == path[0]:
      return

func findPath(nNodes: int; adj: Adj; vertexSet: seq[int]): seq[int] =
  if vertexSet.len == 0:
    return @[]
  let vs = toHashSet(vertexSet)
  let head = vertexSet[0]
  var neighbors: seq[int]
  for w in adj[head]:
    if w in vs:
      neighbors.add w
  if neighbors.len == 0:
    return @[]
  if neighbors.len == 1:
    var path = @[head, neighbors[0]]
    tracePath(nNodes, adj, path, vs)
    return path
  var c0 = @[head, neighbors[0]]
  tracePath(nNodes, adj, c0, vs)
  if c0[^1] == head:
    return c0
  var c1 = @[head, neighbors[1]]
  tracePath(nNodes, adj, c1, vs)
  result = reversed(c0) & c1[1..^1]

# ---------- topology: _divide, noodlize ----------
proc divide(nNodes: int; adj: var Adj; vertex, offset: int) =
  var nei: seq[int]
  for x in adj[vertex]:
    nei.add x
  while nei.len < 4:
    nei.add -1
  nei.setLen(4)
  var valid: seq[int]
  for x in nei:
    if x >= 0:
      valid.add x
  if valid.len < 2:
    return
  shuffle(valid)
  let migrants = [valid[0], valid[1]]
  let newVertex = vertex + offset
  for m in migrants:
    var p = adj[m].find(vertex)
    if p >= 0:
      adj[m].delete(p)
    p = adj[vertex].find(m)
    if p >= 0:
      adj[vertex].delete(p)
    adj[newVertex].add m
    adj[m].add newVertex

## noodlize: グラフを 2*nOrig ノードに拡張するため、ピーク時は元の adj と adjNoodles の両方が生きる（8M 頂点で 2〜3 GB 程度になる要因）
proc noodlize(nOrig: int; adj: Adj; fixedOut, fixedIn: Adj): (int, Adj) =
  let nNodes = 2 * nOrig
  var adjNoodles: Adj
  for v in 0 ..< nOrig:
    adjNoodles.add adj[v].mapIt(it)
  for v in 0 ..< nOrig:
    adjNoodles.add newSeq[int]()
  for u in 0 ..< nOrig:
    let iu = nodeToIdx(u, nOrig)
    for v in fixedOut[iu]:
      if v >= 0 and v < nOrig and v in adjNoodles[u]:
        var p = adjNoodles[u].find(v)
        adjNoodles[u].delete(p)
        p = adjNoodles[v].find(u)
        adjNoodles[v].delete(p)
  for v in 0 ..< nOrig:
    let nfixed = fixedOut[nodeToIdx(v, nOrig)].len + fixedIn[nodeToIdx(v, nOrig)].len
    if nfixed == 0:
      divide(nNodes, adjNoodles, v, nOrig)
  return (nNodes, adjNoodles)

# ---------- topology: _decompose_complex_path ----------
func decomposeComplexPath(path: seq[int]): seq[seq[int]] =
  if path.len == 0:
    return @[]
  var order: Table[int, int]
  var store: seq[int]
  order[path[0]] = 0
  store.add path[0]
  var headp = 1
  while headp < path.len:
    let node = path[headp]
    if node in order:
      let size = order.len - order[node]
      var cycle: seq[int]
      for i in (store.len - size) ..< store.len:
        cycle.add store[i]
      cycle.add node
      result.add cycle
      for v in cycle[1..^1]:
        order.del(v)
      store.setLen(store.len - size)
    order[node] = order.len
    store.add node
    headp += 1
  if store.len > 1:
    result.add store

# ---------- topology: split_into_simple_paths ----------
proc splitIntoSimplePaths(nOrig, nNodes: int; adj: Adj): seq[seq[int]] =
  let components = connectedComponents(nNodes, adj)
  for vertexSet in components:
    let path = findPath(nNodes, adj, vertexSet)
    var flatten: seq[int]
    for v in path:
      flatten.add v mod nOrig
    for p in decomposeComplexPath(flatten):
      result.add p

# ---------- topology: connect_matching_paths 用ヘルパ ----------
func copyDirected(nOrig: int; outAdj, inAdj: Adj): (Adj, Adj) =
  let size = nOrig + 4
  var oa, ia: Adj
  for i in 0 ..< size:
    oa.add outAdj[i].mapIt(it)
    ia.add inAdj[i].mapIt(it)
  return (oa, ia)

func getPerimeters(nOrig: int; adj, outAdj, inAdj: Adj): (HashSet[int], HashSet[int]) =
  var inPeri, outPeri: HashSet[int]
  for i in 0 ..< nOrig:
    let degG = adj[i].len
    let inD = inAdj[i].len
    let outD = outAdj[i].len
    if inD + outD >= degG:
      continue
    if inD > outD:
      outPeri.incl i
    elif inD < outD:
      inPeri.incl i
  return (inPeri, outPeri)

proc chooseFreeEdge(nOrig: int; adj, outAdj, inAdj: Adj; node: int): int =
  var neis: seq[int]
  for x in adj[node]:
    neis.add x
  while neis.len < 4:
    neis.add -1 - neis.len
  neis.setLen(4)
  shuffle(neis)
  for nei in neis:
    if nei == -1:
      continue
    let iNode = nodeToIdx(node, nOrig)
    let iNei = if nei >= 0: nodeToIdx(nei, nOrig) else: nOrig + (-1 - nei)
    let hasOut = nei in outAdj[iNode]
    let hasIn = node in outAdj[iNei]
    if not (hasOut or hasIn):
      return nei
  return -999

proc removeDummyNodes(nOrig: int; outAdj, inAdj: var Adj) =
  let size = nOrig + 4
  for i in nOrig ..< size:
    outAdj[i].setLen(0)
    inAdj[i].setLen(0)
  for v in 0 ..< nOrig:
    outAdj[v] = outAdj[v].filterIt(it >= 0 and it < nOrig)
    inAdj[v] = inAdj[v].filterIt(it >= 0 and it < nOrig)

proc connectMatchingPaths(nOrig: int; adj, fixedOut, fixedIn: Adj): Option[(Adj, Adj, seq[seq[int]])] =
  var (fOut, fIn) = copyDirected(nOrig, fixedOut, fixedIn)
  var (inPeri, outPeri) = getPerimeters(nOrig, adj, fOut, fIn)
  var derivedCycles: seq[seq[int]]

  proc addEdge(u, v: int) =
    let iu = nodeToIdx(u, nOrig)
    let iv = nodeToIdx(v, nOrig)
    fOut[iu].add v
    fIn[iv].add u

  func inDegree(node: int): int = fIn[nodeToIdx(node, nOrig)].len
  func outDegree(node: int): int = fOut[nodeToIdx(node, nOrig)].len

  while outPeri.len > 0:
    var node = 0
    for n in outPeri:
      node = n
      break
    outPeri.excl node
    var path = @[node]
    var cur = node
    block innerOut:
      while true:
        if cur < 0:
          break innerOut
        if cur in inPeri:
          inPeri.excl cur
          break innerOut
        if outDegree(cur) * 2 > 4:
          return options.none[(Adj, Adj, seq[seq[int]])]()
        if adj[cur].len == outDegree(cur) + inDegree(cur):
          return options.none[(Adj, Adj, seq[seq[int]])]()
        let nextNode = chooseFreeEdge(nOrig, adj, fOut, fIn, cur)
        if nextNode == -999:
          return options.none[(Adj, Adj, seq[seq[int]])]()
        addEdge(cur, nextNode)
        if nextNode >= 0:
          path.add nextNode
          if inDegree(cur) > outDegree(cur):
            outPeri.incl cur
        cur = nextNode
        for i in 0 ..< path.len - 1:
          if path[i] == cur:
            derivedCycles.add path[i..^1]
            path.setLen(i + 1)
            break

  while inPeri.len > 0:
    var node = 0
    for n in inPeri:
      node = n
      break
    inPeri.excl node
    var path = @[node]
    var cur = node
    block innerIn:
      while true:
        if cur < 0:
          break innerIn
        if cur in outPeri:
          outPeri.excl cur
          break innerIn
        if outDegree(cur) * 2 > 4:
          return options.none[(Adj, Adj, seq[seq[int]])]()
        if adj[cur].len == outDegree(cur) + inDegree(cur):
          return options.none[(Adj, Adj, seq[seq[int]])]()
        let nextNode = chooseFreeEdge(nOrig, adj, fOut, fIn, cur)
        if nextNode == -999:
          return options.none[(Adj, Adj, seq[seq[int]])]()
        if nextNode >= 0:
          path.add nextNode
        addEdge(nextNode, cur)
        if nextNode >= 0 and inDegree(cur) < outDegree(cur):
          inPeri.incl cur
        cur = nextNode
        for i in 0 ..< path.len - 1:
          if path[i] == cur:
            derivedCycles.add path[i..^1]
            path.setLen(i + 1)
            break

  removeDummyNodes(nOrig, fOut, fIn)
  return some((fOut, fIn, derivedCycles))

# ---------- directed edges from arrays ----------
func arraysToDirectedEdges(nOrig: int; outAdj, inAdj: Adj): DirectedEdges =
  for i in 0 ..< outAdj.len:
    let u = idxToNode(i, nOrig)
    if u < 0:
      continue
    for v in outAdj[i]:
      if v >= 0 and v < nOrig:
        result.add (u, v)

func pathEdges(paths: seq[seq[int]]): DirectedEdges =
  for path in paths:
    for i in 0 ..< path.len - 1:
      result.add (path[i], path[i+1])

func finallyFixedEdges(nOrig: int; fixedOut, fixedIn: Adj; derivedCycles: seq[seq[int]]): DirectedEdges =
  let edges = arraysToDirectedEdges(nOrig, fixedOut, fixedIn)
  var cycleSet: HashSet[(int,int)]
  for cycle in derivedCycles:
    for i in 0 ..< cycle.len - 1:
      cycleSet.incl (cycle[i], cycle[i+1])
  for (u, v) in edges:
    if (u, v) notin cycleSet:
      result.add (u, v)

# ---------- dipole (省略版: positions なしなら optimize は identity) ----------
# ここでは dipole 最適化は未実装とする。paths をそのまま使う。

# ---------- main: iceGraphEdges ----------
proc iceGraphEdges*(
  nOrig: int,
  adj: Adj,
  fixedOut, fixedIn: Adj,
  pairingAttempts: int = 100,
): Option[DirectedEdges] =
  ## 固定辺なしの場合は fixedOut/fixedIn は size nOrig+4 の空 seq を渡す。
  var fOut = fixedOut
  var fIn = fixedIn
  var derivedCycles: seq[seq[int]] = @[]

  let nFixed = block:
    var c = 0
    for i in 0 ..< fOut.len:
      c += fOut[i].len
    c

  if nFixed > 0:
    block tryPairing:
      for attempt in 0 ..< pairingAttempts:
        let res = connectMatchingPaths(nOrig, adj, fOut, fIn)
        if res.isSome:
          let (o, i, dc) = res.get
          fOut = o
          fIn = i
          derivedCycles = dc
          break tryPairing
      return options.none[DirectedEdges]()

  let finFixed = finallyFixedEdges(nOrig, fOut, fIn, derivedCycles)
  let (nNodes, adjNoodles) = noodlize(nOrig, adj, fOut, fIn)
  var paths = splitIntoSimplePaths(nOrig, nNodes, adjNoodles)
  for c in derivedCycles:
    paths.add c

  var allEdges = finFixed
  allEdges.add pathEdges(paths)
  return some(allEdges)

# ---------- 使用例 ----------
when isMainModule:
  randomize()
  let adj: Adj = @[
    @[1, 2],
    @[0, 2],
    @[0, 1],
    @[4],
    @[3],
  ]
  let nOrig = 5
  var fixedOut: Adj
  for i in 0 ..< nOrig + 4:
    fixedOut.add newSeq[int]()
  var fixedIn: Adj
  for i in 0 ..< nOrig + 4:
    fixedIn.add newSeq[int]()

  let res = iceGraphEdges(nOrig, adj, fixedOut, fixedIn)
  if res.isSome:
    echo "edges: ", res.get.len
    echo "sample: ", res.get[0..min(4, res.get.len-1)]
  else:
    echo "no solution"
